appinventor/buildserver/src/com/google/appinventor/buildserver/util/Execution.java

```diff
@@ -132,10 +145,17 @@ public static int execute(File workingDir, String[] command, StringBuffer out,
       StringBuffer err) throws IOException {
     LOG.log(Level.INFO, "____Executing " + joiner.join(command));
     Process process = Runtime.getRuntime().exec(command, null, workingDir);
+    // Prevent any interactive shell from waiting for input
+    process.getOutputStream().close();
     Thread outThread = new RedirectStreamToStringBuffer(out, process.getInputStream());
     Thread errThread = new RedirectStreamToStringBuffer(err, process.getErrorStream());
+
     try {
-      process.waitFor();
+      if (!process.waitFor(SUBPROCESS_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
+        process.destroyForcibly();
+        err.append("Process had to be forcibly terminated due to timeout\n");
```

Comment:
Shouldn't we return here?

Reply:
We have to first join the threads below, no? Otherwise they will keep running forever. Basically, it waits, forces stop, joins threads, and finally it returns.

Reply:
Yes I think you are right but this feels really unintuitive to me. If the process has failed and then is killed, the IO streams should also close and the threads should terminate. If that's not the case it would be good to understand why.

Reply:
Tbh I was thinking the same, I actually first had a return in case of timeout, but then realized about the hanging threads. In the end, the final result is the same, just some weird code structure...

I just didn't have time to investigate and see what happens with those threads. I'll check later along with your comments on the timeouts above.

Reply:
On a different thought, this function is not used anymore. I guess I'll just remove it 🙂

Reply:
And it's gone now